[2 ~ 4단계 방탈출 결제 / 배포] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4439

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
안녕하세요 파랑~
이번 단계 미션 리뷰도 잘부탁드립니다!😊

#### 미션 관련 url
[배포 서버](http://13.124.20.46:8080/)
[API 명세](http://13.124.20.46:8080/swagger-ui/index.html)
[ERD](https://www.erdcloud.com/d/YjXLNppnG6grnKNrK)

#### 배포 스크립트
```
#!/bin/bash

APP_NAME="spring-roomescape-payment"
VERSION="0.0.1-SNAPSHOT"
REPO_URL="https://github.com/kjyyjk/$APP_NAME.git"
REPO_DIR="/home/ubuntu/$APP_NAME"
JAR_NAME="$APP_NAME-$VERSION"
BUILD_JAR="$REPO_DIR/build/libs/$JAR_NAME.jar"
STDOUT_LOG="$REPO_DIR/build/libs/stdout.log"
STDERR_LOG="$REPO_DIR/build/libs/stderr.log"
BRANCH="step2"

echo "> 배포 시작"

# 프로젝트가 없으면 clone
if [ ! -d "$REPO_DIR" ]; then
  echo "> git clone"
  git clone "$REPO_URL" "$REPO_DIR"
fi

cd "$REPO_DIR"

# 브랜치 checkout 및 최신화
git fetch
git checkout "$BRANCH"
git pull origin "$BRANCH"

# 실행 중인 프로세스 종료 (8080)
PID=$(lsof -ti:8080)
if [ -n "$PID" ]; then
  echo "> 8080 포트 점유 중인 프로세스 종료: PID=$PID"
  kill -9 "$PID"
fi

# 애플리케이션 빌드
echo "> 애플리케이션 빌드 시작"
./gradlew bootJar

# 애플리케이션 실행
echo "> 애플리케이션 실행"
nohup java -jar "$BUILD_JAR" > "$STDOUT_LOG" 2> "$STDERR_LOG" &
```

#### 애플리케이션 실행 관련
애플리케이션 실행 시 사용할 수 있는 admin과 user 계정을 등록해두었습니다. 로그인 정보는 다음과 같습니다.
```
admin
id : admin@email.com
password : qwe123

user1
id : user1@email.com
password : qwe123

user2
id : user2@email.com
password : qwe123
```

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kjyyjk committed:
refactor: yml 설정값을 ConfigurationProperties로 다루도록 리팩터링

kjyyjk committed:
refactor: TossPaymentClient 리팩토링 및 설정 파일 정리

- 특정 외부 API 별 RestClient.Builder를 빈으로 등록한다.
Client에서 특정 Builder를 주입받을 때는 생성자 파라미터명으로 구분한다.

kjyyjk committed:
refactor: 2단게 프론트엔드 파일 수정

kjyyjk committed:
refactor: `@Qualifier`로 주입받는 빈 명시적 설정

kjyyjk committed:
feat: 예약 시 결제 및 결제 내역 저장

- 예약 생성 혹은 결제 승인에 실패하면 어떤 것도 저장하지 않는다.
예약 생성 및 결제 승인에 성공했지만 결제 내역 저장에 실패한 경우에는
로깅만 하고 롤백하지 않는다.

- 결제 승인 외부 api 호출은 커넥션 풀 고갈 문제를 고려해 트랜잭션 외부에서 진행한다.
따라서 결제 승인 외부 api 호출 전후로 존재하는 예약 생성과 결제 저장은 독립적인
트랜잭션에서 진행한다.

kjyyjk committed:
refactor: PaymentService 테스트 수정

kjyyjk committed:
feat: 내 예약 조회 시 결제 내역도 함께 조회

kjyyjk committed:
feat: 로깅 구현

kjyyjk committed:
refactor: 로깅 구현에 따른 테스트 수정

kjyyjk committed:
feat: api 명세

kjyyjk committed:
refactor: swagger에 서버 url 기재

kjyyjk committed:
refactor: IOException을 PaymentException으로 변환

kjyyjk commented on build.gradle:
```diff
@@ -22,15 +22,12 @@ dependencies {
-
     implementation 'org.springframework.security:spring-security-crypto'
-
+    implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.8.8'
```
미션 시간이 빠듯했던 탓에 spring doc과 rest docs 중 spring doc을 활용한 API 명세를 선택했습니다.

spring doc(swagger)는 어노테이션을 기반으로 간편하고 빠르게 문서화할 수 있다는 장점이 있었습니다.
또한 클라이언트 개발자가 문서에서 API 요청을 보내볼 수 있다는 이점도 존재했습니다.

반면 컨트롤러에 어노테이션 기반으로 명세를 하기 때문에, 프로덕션 코드가 지저분해진다는 단점이 있었습니다.

이에 대해서는 Controller 인터페이스를 분리하고, 인터페이스에 swagger 어노테이션을 달아주는 방식으로 가독성의
단점을 보완하고자 했습니다.

summerlunaa replied:
> 반면 컨트롤러에 어노테이션 기반으로 명세를 하기 때문에, 프로덕션 코드가 지저분해진다는 단점이 있었습니다.

저도 회사에서 스웨거를 사용하고 있지만 코드가 지저분해진다는 느낌은 안 들더라구요. 오히려 코드를 보면서 한글로된 구체적인 설명을 함께 볼 수 있으니 이게 어떤 API인지 더 명확히 알 수 있어서 좋았습니다. 저는 swagger를 위한 설명이 추가되는 걸 장점으로 느껴서 그런지 오히려 인터페이스를 분리한 게 더 번거로워 보이긴 하네요 :) 개인적인 의견입니다.

kjyyjk replied:
맞아요... 매우 번거로웠습니다.ㅎ
말씀 주신 것처럼 코드에 스웨거 어노테이션이 추가되는 것을 오히려 장점으로 바라볼 수도 있겠네요!

파랑의 의견을 말씀해주셔서 감사합니다👍

kjyyjk commented on src/main/java/roomescape/common/config/RestClientConfig.java:
```diff
@@ -10,16 +10,12 @@
-    public Builder restClientBuilder() {
+    @Qualifier("tossRestClientBuilder")
+    public Builder tossRestClientBuilder(TossPaymentsProperties tossPaymentsProperties) {
         final SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
-        factory.setConnectTimeout(TIMEOUT_CONNECTION);
-        factory.setReadTimeout(TIMEOUT_READ);
+        factory.setConnectTimeout(tossPaymentsProperties.getConnectionTimeout());
+        factory.setReadTimeout(tossPaymentsProperties.getReadTimeout());
         return RestClient.builder().requestFactory(factory);
     }
```
> 모든 client가 동일한 timeout 값을 사용할까요? 고민해보세요 :)

제가 2초/8초로 timeout을 설정한 것은 토스 클라이언트에 대한 것이지, 다른 클라이언트에 해당하는 것은 아니었습니다.
따라서 tossRestClientBuilder라는 이름을 가진 Builder에 토스 관련 설정을 해주고 빈으로 등록했습니다.

만약 다른 Client를 사용할 일이 있다면 위와 같은 방식으로 해당 RestClient.Builder를 만들어주고
주입 받는 곳에서 특정 빈을 명시해 주입 받게 할 수 있습니다.

kjyyjk commented on src/main/java/roomescape/common/exception/GlobalExceptionHandler.java:
```diff
@@ -3,52 +3,67 @@
         return ResponseEntity.internalServerError().body(errorResponse);
     }
+
+    private void logExceptionToWarn(Exception e) {
+        log.atWarn().log(e.getMessage(), e);
+    }
+
+    private void logExceptionToError(Exception e) {
+        log.atError().log(e.getMessage(), e);
+    }
```
`GlobalExceptionHandler`에서 다루는 모든 예외에 대해 로깅하도록 구현했습니다.

예외에 대한 로그 레벨은 warn과 error로 나누었습니다.
개발자가 휴가 중 연락을 받아야만 할 예외일 경우 error로, 그렇지 않은 경우 warn으로 판단했습니다.😂

kjyyjk commented on src/main/java/roomescape/common/interceptor/LoggingInterceptor.java:
```diff
@@ -0,0 +1,26 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+@Slf4j
+public class LoggingInterceptor implements HandlerInterceptor {
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+            throws Exception {
+        request.setAttribute("uuid", randomUUID().toString());
+        log.atInfo().log("request : uuid={}, path={}", request.getAttribute("uuid"), request.getRequestURI());
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex)
+            throws Exception {
+        log.atInfo().log("response : uuid={}, path={}", request.getAttribute("uuid"), request.getRequestURI());
+    }
+}
```
애플리케이션에 들어오는 모든 요청과 응답에 대해 로깅하도록 구현했습니다.
이때 UUID를 활용해 요청과 그에 대한 응답을 매핑해주었습니다.

현재는 해당 인터셉터에서 수행하는 요청/응답 로깅에 대해서만 UUID를 구분하고 있습니다.
이후에는 다른 비즈니스 로직의 로그에서도 UUID를 포함해주면 좋겠다고 생각했습니다.
다만 비즈니스 로직에서 어떻게 request#getAttribute()를 할 수 있을지에 대해서는 더 고민해봐야할 것 같습니다..!

summerlunaa replied:
UUID까지 고민해주셨네요! 당장 적용하진 않는 걸 추천드리지만..! 궁금하시다면 아래 키워드에 대해 공부해보세요.
- MDC (Mapped Diagnostic Context)
- 로깅 설정 파일에서의 로그 패턴 설정

추가로 요청 객체나 파라미터, 응답 객체 등은 로깅 필요 없을까요? 어떤 정보가 추가되면 더 유용한 로그가 될 수 있을까요?

kjyyjk replied:
> 당장 적용하진 않는 걸 추천드리지만..!

추천하지 않는다고 하신 것이 UUID 자체에 대한 내용일까요? 아니면 제시해주신 키워드에 대한 내용일까요?

kjyyjk replied:
> 요청 객체나 파라미터, 응답 객체 등은 로깅 필요 없을까요? 어떤 정보가 추가되면 더 유용한 로그가 될 수 있을까요?

제가 막연하게 생각해본 로깅의 목적은 아래 세가지 입니다.
- 에러 추적
- 사용자 요청 통계 및 분석
- 잘 동작하는지 흐름 관찰

이러한 목적을 잘 달성하기 위해 로깅에 추가하면 좋을 것 같은 정보들은 다음과 같습니다.
- Http Method (uri path 만으로는 어떤 요청인지 구분이 불가하니까)
- 요청 객체, 응답 객체 (비즈니스 로직과는 별개로 어떤 요청이 들어오고 어떤 응답이 나갔는지를 기록)
- 헤더 정보

summerlunaa replied:
> 추천하지 않는다고 하신 것이 UUID 자체에 대한 내용일까요? 아니면 제시해주신 키워드에 대한 내용일까요?

키워드에 대한 내용입니다. 그렇게까지 중요한 부분은 아닌데 찾아보고 적용하는데 시간 많이 쓰실까봐 그렇게 말씀 드렸습니다 ㅎㅎ

kjyyjk commented on src/main/java/roomescape/payment/domain/PaymentHistory.java:
```diff
@@ -0,0 +1,42 @@
+    @Enumerated(EnumType.STRING)
+    private PaymentStatus status;
+    @JoinColumn
+    @OneToOne
+    private Reservation reservation;
```
PaymentHistory가 Reservation에 대한 외래키를 가지도록 설계했습니다.

사실 누가 외래키를 가져도 크게 문제가 되지 않을 것 같아서 한 2일정도 고민했습니다.
그 과정에서 많은 크루들의 생각을 들어보고 같이 의논해보기도 했는데요!

제 결론은 다음과 같습니다.

현재는 PaymentHistory와 Reservation이 일대일 관계라서 고민하고 있지만, 만약 일대다(혹은 다대일) 관계였다면
무조건 `다` 쪽에 외래키를 두어야합니다.

현재 기준이 모호하므로 추후 확장성을 고려해서 `다`가 될 것 같은 쪽에 외래키를 위치하는게 좋다고 판단했습니다.
상상해봤을 때 하나의 결제로 여러 예약을 할 수 있지는 않을 것 같고, 하나의 예약에 대해서는 여러 결제 내역이 생길 수 있을 것 같습니다.

따라서 기능 확장으로 인해 `다`가 될 수 있는 PaymentHistory에 외래키를 위치했습니다.

summerlunaa replied:
왜 Payment가 아니라 PaymentHistory인가요? history라는 단어가 붙었다는 건 `결제한 이력을 저장하겠다`라는 느낌이거든요. `이력`이라는 건 특별한 이유가 없다면 데이터가 변경되거나 삭제될 일이 없고 계속해서 쌓이는 데이터라는 뜻이구요. 이런 의미의 도메인을 의도하신 건가요?
이력은 일반적으로 데이터가 삭제되어서도 안 되고 변경되어서도 안 되는데 다른 테이블과의 연관관계가 있으면 어떤 문제가 있을까요? 예를 들어 주문 내역을 생각해보세요. 내가 '맛있는 사과'를 1만원에 구매했다면 주문 내역에는 '맛있는 사과', 1만원이 기록되어야 하고 이 내용은 변해서는 안 됩니다. 그런데 주문 내역 테이블이 상품 테이블을 참조한다면 어떤 일이 벌어질까요? 주문 이후 상품이 '아주 맛있는 사과', 8천원으로 변경되는 경우 나의 주문 내역까지 변경될 거예요. `history` 데이터를 다룰 때에는 이런 부분들까지 고려해야 합니다 ㅎㅎ
또 이력을 통해 데이터의 변경 흐름도 파악할 수 있어야 하기 때문에 데이터가 변경되는 경우에도 history는 계속 쌓여야 합니다. 그래야 `6/10 00:00`에 결제가 완료되었다가 `6/11 12:00`에 결제가 취소되었다는 이력을 알 수 있으니까요. 이 경우 변경되는 건 PaymentHistory가 아니가 Payment여야겠죠. 이런 부분들까지 고민해보세욥.

kjyyjk replied:
제가 의도한 `PaymentHistory`는 결제 완료에서 결제 취소로 상태가 수정될 수 있습니다.
하지만 말씀해 주신 것처럼 `History`라면 취소 됐다고 취소 상태로 변할 것이 아니라, 취소 내역이 추가되어야 할 것 같아요.
그런 의미에서 `PaymentHistory`보다는 `Payment`라는 표현이 더 적절하다고 판단해 수정했습니다!
8f1635806d61a42ee7fdf2d4f2c7fe61063e9f55

만약 정말 내역의 의미로 `History`를 추가하게 된다면 다른 엔티티와 연관 관계를 추가하지 않을 것 같습니다.
내역이기 때문에 그 내용이 변하면 안될 뿐더러 cascade와 같은 옵션으로 인해 잘못 조작될 가능성도 있기 때문입니다.

kjyyjk commented on src/main/java/roomescape/reservation/log/ReservationProbe.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.stereotype.Component;
+import roomescape.reservation.domain.Reservation;
+
+@Slf4j
+@Component
+public class ReservationProbe {
+
+    public void create(Reservation reservation) {
+        log.atInfo().log("create reservation: id={}, date={}, theme={}, time={}, reservationMemberId={}",
+                reservation.getId(), reservation.getDate().getValue(), reservation.getTheme().getName().getValue(),
+                reservation.getTime().getStartAt(), reservation.getMember().getId());
+    }
+
+    public void delete(Reservation reservation) {
+        log.atInfo().log("delete reservation: id={}, date={}, theme={}, time={}, reservationMemberId={}",
+                reservation.getId(), reservation.getDate().getValue(), reservation.getTheme().getName().getValue(),
+                reservation.getTime().getStartAt(), reservation.getMember().getId());
+    }
+}
```
이와 같이 특정 도메인에 대한 로깅의 책임은 Probe라는 객체를 분리해 위임했습니다.

구체적인 로깅 구현을 캡슐화하여 유지보수에 좋다는 이점이 있습니다.
하지만 객체 지향의 단점과 마찬가지로 클래스 수가 늘고 복잡도가 증가한다는 단점이 있습니다.

저는 단점보다는 유지 보수성으로 얻는 비용 절감의 효과가 더 크다고 판단해 해당 방식을 택했습니다.

kjyyjk replied:
추가로 내부에서 발생하는 특정 중복 로직도 제거해볼 수 있을 것 같습니다.

summerlunaa replied:
로깅 인터셉터에서 요청과 응답 객체를 로깅하면 요런 건 필요 없을 것 같은데 어떻게 생각하시나욥 ㅎㅎ

kjyyjk replied:
해당 코드는 예약 생성 및 저장과 같은 비즈니스 로직에 대한 로깅을 담당하고 있습니다.
반면 인터셉터에서의 로깅은 들어오고 나가는 요청/응답에만 초점이 맞춰져 있다고 생각합니다.

현재로써는 주요 비즈니스 로직에 대한 로깅이 있어야만 내부적으로 어떤 일들이 발생했는지를 디버깅할 수 있다고 생각하는데
어떻게 생각하시나요?

인텔리제이 같은 idea를 통해 디버깅을 할 때 시작과 끝만이 아닌 중간 과정들을 살펴 보듯이,
로깅의 목적인 운영 환경 디버깅을 위해서는 주요 비즈니스 로직 동작에 대한 로그들도 필요하다고 생각됩니다!

summerlunaa replied:
요청/응답으로도 충분히 파악할 수 있는 부분이라 생각해서 코멘트 드렸습니다! 요청/응답과 위 로그 모두 남게 되면 로그가 거의 2배가 되는데요 로그 데이터 양이 너무 많아져도 로그 파악이나 로그 데이터 관리가 쉽지 않아서요. 요청/응답 로그가 이미 존재하는데도 모든 생성 요청과 삭제 요청에 대해 위 내용을 기록할 필요가 있는지 고민해보면 좋을 것 같습니다 :)

kjyyjk commented on src/main/java/roomescape/reservation/service/ReservationPayService.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private void savePaymentHistory(CreateReservationWithPaymentWebRequest webRequest, Reservation savedReservation) {
+        try {
+            PaymentHistory paymentHistory = webRequest.toPaymentHistory(savedReservation, PaymentStatus.DONE);
+            paymentHistoryRepository.save(paymentHistory);
+        } catch (IllegalArgumentException | OptimisticLockException e) {
+            log.atError().log("결제 내역 DB 저장 실패 - 사용자에게는 정상 응답. "
+                            + "paymentKey:{}, orderId:{}, amount:{}, reservationId:{}",
+                    webRequest.paymentKey(), webRequest.orderId(), webRequest.amount(), savedReservation.getId()
+            );
+        }
+    }
```
예약 생성과 결제 승인이 모두 완료되었지만 결제 내역 저장에 실패한 경우 롤백하지 않고 정상 응답하게끔 구현했습니다.
대신 결제 내역이 누락되고 있는 중요한 사안인만큼 error 레벨로 로깅을 해주었습니다.

이처럼 서버 개발자가 조금 고생하면..?🤔
별도로 결제 승인 취소를 하지 않아도 되고, 사용자의 편의성도 증가할 것이라고 기대했습니다!

summerlunaa replied:
어떤 사유로 에러가 발생했는지는 로깅하지 않아도 될까요?

kjyyjk replied:
현재 정보로는 문제가 `왜` 발생했는지 디버깅하기 힘들 것 같아요.
따라서 e에 대한 스택 트레이스를 포함시켰습니다!
95a1838f4084e2fc86460f18a210078f4bc649de

kjyyjk commented on src/main/resources/application.yml:
```diff
@@ -25,3 +22,19 @@ spring:
   sql:
     init:
       mode: always
+
+jwt:
+  key: /VCEkdrvbjkCLRp0G0LK12VsDYViO/kufd8UVphCq80MdiZzNlefqOcwqfpvW+jIRNBmhiNKbNUGXSwum0DSgUjkEl2To/350R/ffncBkVeO7tfNq/4rNXWGGHre9p8s14m0k2z/QS4=
+  expiration: 360000
```
> 그런데 두 yml을 하나로 합쳐주실 수 있을까요?

파랑의 편의를 위해서 망설임 1도 없이 합쳤습니다ㅎㅎ

kjyyjk reviewed (commented):
레벨 2의 진짜진짜 마지막 단계네요..!
잘부탁드립니다.👍

+서버가 잘 죽는 것 같아서,, 혹시 배포 서버나 API 명세에 접근이 불가하시다면 말씀주세요! 바로 조치하겠습니다.

kjyyjk commented on src/main/java/roomescape/common/config/TossPaymentsProperties.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@Getter
+@RequiredArgsConstructor
+@ConfigurationProperties(prefix = "pay.toss")
+public class TossPaymentsProperties {
+    private final String secretKey;
+    private final String baseUrl;
+    private final TimeoutProperties timeout;
+
+    public int getConnectionTimeout() {
+        return timeout.getConnection();
+    }
+
+    public int getReadTimeout() {
+        return timeout.getRead();
+    }
+}
```
> 참고) `@ConfigurationProperties`와 `@ConfigurationPropertiesScan`을 사용해서 설정값을 가져올 수도 있습니다.

말씀해주신 `@ConfigurationProperties`에 대해 학습해보고 적용해보았습니다.
당장에 필요성을 느끼지 못하고 적용했지만 사용하면서 느꼈던 이점은 다음과 같습니다.

- 불변성을 보장해줄 수 있다.
- 간편하게 빈 주입 받아 사용할 수 있다.
- 객체 그래프로 구성될 수 있어 여러 곳에서 재사용될 수 있다.
- properties나 yml과 같은 설정 파일에 대한 결합도가 줄어, 테스트 시에 용이하다.

summerlunaa commented on src/main/java/roomescape/common/config/SpringDocsConfig.java:
```diff
@@ -0,0 +1,24 @@
+        info = @Info(title = "방탈출 API", version = "v1",
+                description = "우아한테크코스 7기 레벨2 방탈출 예약 관리 시스템",
+                contact = @Contact(name = "레오", email = "rlawnsdud920@gmail.com")),
+        servers = @Server(url = "http://3.35.134.88:8080/")
```
스웨거의 서버 URL이 하드코딩 되어있네요. 어떤 문제가 생길 수 있을까요? 어떻게 개선할 수 있을까요?
그리고 배포된 서비스의 URL과 스웨거의 URL이 다르네요. 스웨거에 API 호출 기능이 있다는 걸 아시나요? 잘 동작하는지 확인해보세요.

kjyyjk replied:
서버 URL이 변경될 때마다 하드코딩 된 스웨거의 서버 URL도 직접 수정해주어야 합니다.
현재는 ec2 인스턴스를 중지/시작하면 URL이 변경되어버리니 현재처럼 불일치 문제가 자주 발생할 것 같아요.
이에 대해서는 탄력적 IP를 할당 받아 서버의 URL이 거의 변경되지 않게끔 조치할 수 있습니다.

`하드코딩`에 초점을 두어 바라본다면 환경 변수/설정 정보로 분리해볼 수도 있을 것 같습니다.
현재처럼 하드 코딩한다면 개발 환경에 따라 다른 URL을 설정하지 못하지만, 분리할 경우 개발 환경에 따라 다른 설정 정보를 적용할 수 있다는 이점이 있습니다.

현재 탄력적 IP 할당 권한이 부여되지 않아 설정 파일로만 분리했습니다.
55839c4556bce622d66abdb25150090071e1bf11

summerlunaa commented on src/main/java/roomescape/common/config/SpringDocsConfig.java:
```diff
@@ -0,0 +1,24 @@
+                contact = @Contact(name = "레오", email = "rlawnsdud920@gmail.com")),
+        servers = @Server(url = "http://3.35.134.88:8080/")
+)
+@SecurityScheme(
+        name = "cookieAuth",
+        type = SecuritySchemeType.APIKEY,
+        in = SecuritySchemeIn.COOKIE,
+        paramName = "access_token"
+)
```
description을 통해 설명을 조금 더 추가해주면 좋을 것 같아요.

kjyyjk replied:
`쿠키에 담긴 JWT 기반 인증 및 인가`라는 설명을 추가해주었습니다.
22045c6db19fdf251651d931058065359185d664

summerlunaa commented on src/main/java/roomescape/member/controller/AuthController.java:
```diff
@@ -1,57 +1,51 @@
+@Tag(name = "로그인 API")
+public interface AuthController {
+
+    @Operation(summary = "로그인")
+    @RequestBody(
+            required = true,
+            content = @Content(
+                    schema = @Schema(implementation = LoginRequest.class),
+                    examples = {
+                            @ExampleObject(name = "sample",
+                                    value = "{\"email\":\"user1@email.com\",\"password\":\"qwe123\"}")
+                    }
+            )
+    )
+    ResponseEntity<Void> login(@RequestBody(required = true) LoginRequest loginRequest);
```
예시까지 잘 적어주셨네요 👍 참고로 요청/응답 DTO의 필드에 `@Schema(description = "XXX")` 애노테이션을 추가하면 해당 필드에 대한 설명이나 예시, 필수 여부 등을 추가로 적어줄 수 있어요. GET API의 파라미터에서는 `@Parameter` 애노테이션으로 추가 가능하구요. 현재는 요청/응답 필드에 대한 설명이나 nullable 여부를 확인할 수 없네요. 더 친절한 API 문서를 만들려면 스웨거에 어떤 애노테이션이 있는지 찾아보고 적용해보는 것도 좋을 것 같습니다 :)

kjyyjk replied:
요청 파라미터에 대한 `@Parameter`, `@Schema`, `@ApiResponse` 등 어노테이션을 활용하면 더 클라이언트 친화적인 명세를
만들어갈 수 있을 것 같습니다.
현재로써는 어떤 것이 있는지만 알아두어도 충분하다고 판단해, 따로 적용은 하지 않았습니다!

summerlunaa commented on src/main/java/roomescape/payment/toss/TossPaymentErrorHandler.java:
```diff
@@ -16,16 +16,20 @@ public class TossPaymentErrorHandler extends DefaultResponseErrorHandler {
+
+                boolean exists = FilteredPaymentErrorCode.exists(tossPaymentError.code());
+                if (exists) {
+                    throw new PaymentException(INTERNAL_SERVER_ERROR, "서버 내부 오류입니다.", INTERNAL_SERVER_ERROR.name());
```
로깅을 도입한 김에 여기와 아래도 로그 챙겨볼까요?

kjyyjk replied:
filtered status code에 해당하는 에러/해당하지 않는 에러/io exception 이 세가지 경우에 대한 로깅 추가했습니다.
ea1733f465b0b414f66e4114c4a8735346917c6e

summerlunaa replied:
INTERNAL_SERVER_ERROR의 경우 토스에서 받은 에러 코드와 메시지를 사용하지 않기 때문에 원인 파악이 힘들 것이라 생각해서 추가 요청 드렸습니다. `FilteredPaymentErrorCode에 해당하지 않는 에러`의 경우 로그가 필요하지 않겠네요.

summerlunaa commented on src/main/java/roomescape/reservation/controller/ReservationAdminController.java:
```diff
@@ -1,62 +1,46 @@
-import roomescape.common.utils.UriFactory;
-import roomescape.member.auth.RoleRequired;
-import roomescape.member.domain.Role;
+import roomescape.member.controller.dto.SignupRequest;
```
미사용 import네요

summerlunaa commented on src/main/java/roomescape/reservation/controller/ReservationAdminControllerImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+    private final ReservationService reservationService;
+
+    @RoleRequired(value = Role.ADMIN)
+    @GetMapping(BASE_PATH)
+    public List<ReservationWebResponse> getAll() {
+        return reservationService.getAll();
+    }
+
+    @RoleRequired(value = Role.ADMIN)
+    @GetMapping(BASE_PATH + "/wait")
+    public List<ReservationWaitWebResponse> getAllReservationWait() {
+        return reservationService.getAllReservationWait();
+    }
```
요 두 개는 왜 admin으로 시작하지 않나유?

kjyyjk replied:
이전 미션의 페어 코드였어서 놓쳤던 부분입니다..!
수정했습니다.
550a837f0d7c2735ba2739ce52963f90610f8624

summerlunaa commented on src/main/java/roomescape/common/interceptor/LoggingInterceptor.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex)
+            throws Exception {
```
```suggestion
            {
```

summerlunaa commented on src/main/java/roomescape/payment/PaymentService.java:
```diff
@@ -11,7 +15,12 @@ public class PaymentService {
+        try {
+            // TODO 요청과 응답의 값 변조 검증
+            return paymentClient.confirm(request);
+        } catch (CancellationException | ResourceAccessException e) {
```
`CancellationException`과 `ResourceAccessException`은 어떤 경우 발생할 수 있는 예외인가요? 여기도 로깅 추가할까요?

kjyyjk replied:
`ResourceAccessException`은 Http 요청 중 TimeOut되는 시나리오에서 발생할 수 있는 예외입니다.
그리고 아래 스프링 프로젝트의 두 이슈를 통해 스프링의 특정 버전에서 `ResourceAccessException` 대신 
`CancellationException`이 발생하는 것을 확인할 수 있었습니다.
[문제 이슈](https://github.com/spring-projects/spring-framework/issues/33973)
[해결 이슈](https://github.com/spring-projects/spring-framework/pull/34721)
따라서 조금 더 안전한 예외 처리를 위해 두 예외 모두 catch 해주었습니다.

catch 구문에서 따로 로깅을 하지는 않았지만, GlobalExceptionHandler에서 PaymentException을 처리하며
스택 트레이스를 포함해 로깅하고 있는 것으로 확인했습니다!

summerlunaa replied:
같은 PaymentException 이라도 500에러면 error로 로그를 남겨야 확인이 쉽지 않을까 싶네욥 ㅎㅎ

summerlunaa commented on src/main/java/roomescape/reservation/controller/dto/ReservationWithStatusResponse.java:
```diff
@@ -20,29 +21,37 @@ public class ReservationWithStatusResponse {
-    public static ReservationWithStatusResponse from(
-            final Reservation reservation
-    ) {
+    public static ReservationWithStatusResponse of(Reservation reservation, PaymentHistory paymentHistory) {
```
```suggestion
    public static ReservationWithStatusResponse of(Reservation reservation, @Nullable PaymentHistory paymentHistory) {
```
애노테이션을 통해 nullable한 값임을 명시해주면 더 안전할 것 같아요.

kjyyjk replied:
null이냐 아니냐에 따라 다르게 동작하기 때문에 명시해주면 가독성도 좋아질 것 같아요!
적용했습니다.
6ad750258abb79c8e7af66ec17f40e64c4b726b6

kjyyjk replied:
수정했습니다!
6ad750258abb79c8e7af66ec17f40e64c4b726b6

summerlunaa commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -37,38 +35,37 @@ public class ReservationService {
+                .map(reservation -> ReservationWithStatusResponse.of(
+                        reservation,
+                        paymentHistoryRepository.findByReservation(reservation).orElse(null))
+                ).toList();
```
예약 개수만큼 결제 조회 쿼리가 나가겠네요. 개선 가능할까요?

kjyyjk replied:
예약과 결제 정보가 함께 필요해서 발생하는 N+1문제인 것 같네요!
Reservation이 양방향으로 Payment를 의존하고 있다면 fetch join으로 간편하게 해결할 수 있을 것 같습니다.
하지만 현재 구조에서는 양방향을 지양하고 있기 때문에 다른 방법을 생각해봤습니다.

첫 번째 방법은 ReservationRepository에서 조회할 때 ReservationWithPayment와 같은 Dto를 조회하는 방식입니다.
이 경우에는 한 번의 쿼리만으로 로직을 수행할 수 있습니다.

두 번째 방법은 하나의 in 쿼리를 통해 n개의 예약에 대한 Payment n개를 조회하는 방식입니다.
이 경우에는 총 두 번의 쿼리로 로직을 수행할 수 있습니다.

쿼리문이 복잡해지는 것을 원하지 않기도 하고, 직접 구현해보고 싶었던 방식이었어서 저는 두 번째를 택했습니다.
f7093f7ba2aabc62ae283c3ffc8c9187b272a247

summerlunaa reviewed (changes requested):
레오 안녕하세요!
2-4단계도 잘 구현해주셨네요. 코드에 코멘트 꼼꼼히 달아주셔서 보기 편했어요!
API 문서화와 로깅도 잘 챙겨주셨구요, 배포된 서비스도 확인했습니다 👍
리뷰 남겨드렸으니 확인해보시고 편하게 질문 주세요!

kjyyjk committed:
refactor: PaymentHistory에서 Payment로 변경

- History는 내역이라는 의미가 강하다.
내역은 결제가 취소 되어도 취소 상태로 변경되기 보다는 취소 내역이 따로 추가되어야 한다.
내가 의도했던 PaymentHistory는 그런 내역의 의미와는 거리가 멀다.
따라서 Payment로 이름을 변경한다.

kjyyjk committed:
refactor: 결제 정보 저장 실패 로깅에 스택 트레이스 추가

kjyyjk committed:
refactor: springdoc의 서버 url 설정 파일로 분리

kjyyjk committed:
refactor: security scheme 설명 추가

kjyyjk committed:
refactor: TossPaymentErrorHandler 로깅 추가

kjyyjk committed:
refactor: `@Nullable`로 null일 수도 있음을 명시

kjyyjk committed:
refactor: 내 예약 및 결제 정보 조회 로직 성능 개선

- 내 예약 조회 api에서 예약에 따른 결제 정보를 조회하며 N+1 문제가 발생하고 있다.
단방향인 관계로 Reservation이 Payment를 의존하지 않고 있어, fetch join으로는 해결할 수 없다.
따라서 in 쿼리 및 매핑을 통해 문제를 해결했다.

kjyyjk committed:
refactor: admin 권한 api url 수정

kjyyjk committed:
fix: 테스트 에러 수정

kjyyjk committed:
refactor: swagger 토큰 파라미터명 수정

kjyyjk committed:
refactor: swagger 보안 스키마명 수정

summerlunaa reviewed (approved):
레오 안녕하세요~
리뷰 반영도, 질문에 대한 답변도 너무 잘 해주셨습니다. 더 궁금한 점 있으면 언제든지 DM 주세요!
드디어 레벨 2 마지막 미션도 끝이네요 🎊 방학동안 푹 쉬면서 충전하시고 남은 우테코 생활도 잘 헤쳐나가시길 바랍니다~~ 너무 수고하셨습니다!!
